engine/grid.py

```diff
@@ -300,73 +307,164 @@ def find_border_nodes(self):
+            """
+            Returns a circle of nodes starting from the start_pos, going at orientation orientation, and ending at the
+            end_pos with center center.
+
+            Arguments:
+                start_pos: float tuple representing the starting point of the circle being plotted
+                end_pos: float tuple representing the ending point of the circle being plotted
+                center: float tuple representing the center point of the circle being plotted
+                orientation: orientation that the nodes are being plotted from, starting with the start_pos and ending
+                    at the end_pos
+            """
+            r = math.hypot(float(start_pos[0]) - center[0], float(start_pos[1]) - center[1])
+            theta_init = math.atan2(start_pos[1]-center[1], start_pos[0]-center[0])
+            theta_end = math.atan2(end_pos[1]-center[1], end_pos[0]-center[0])
+            circle_plt = []
+            if theta_end < theta_init:
+                theta_end = theta_end + 2*math.pi
+            if orientation == Orientation.CCW:
+                theta = theta_init
+                while theta < theta_end:
+                    circle_plt.append((r*math.cos(theta)+center[0], r*math.sin(theta)+center[1]))
+                    theta = theta + math.pi/12
+            elif orientation == Orientation.CW:
+                theta = theta_end
+                while theta > theta_init:
+                    circle_plt.append((r*math.cos(theta)+center[0], r*math.sin(theta)+center[1]))
+                    theta = theta - math.pi/12
+            else:
+                raise Exception("invalid orientation")
+            return circle_plt
+
         phase = WaypointPhase.DOWN
         curr_pos = self.leftmost_node_pos
+        direction = Direction.RIGHT
         waypoints = []
         waypoints.append(curr_pos)
         while (phase != WaypointPhase.TERMINATE):
-            if (phase == WaypointPhase.DOWN):
-                new_pos = (curr_pos[0]+1, curr_pos[1])
-                if curr_pos[0]+1 == rows and self.nodes[new_pos].is_active:
+            if direction == Direction.LEFT:
+                new_pos = (curr_pos[0]-1,curr_pos[1])
+                if self.nodes[new_pos].is_active:
                     waypoints.append(new_pos)
-                    right_pos = self.bottom_rightmost_node(new_pos)
-                    if right_pos is not None:
-                        waypoints.append(right_pos)
-                        curr_pos = right_pos
-                        phase = WaypointPhase.DOWN
+                    curr_pos = new_pos
+                else:
+                    left_pos = self.bottom_rightmost_node(new_pos)
+                    if left_pos is not None:
+                        new_pos = (new_pos[0]+1, new_pos[1])
+                        while new_pos[0] < left_pos[0]:
+                            while new_pos in waypoints:
+                                waypoints.remove(new_pos)
```

Comment:
isn't this removing your own progress? On line 423, isn't new_pos curr_pos (since we have new_pos = (curr_pos[0]-1,curr_pos[1]) on line 416). In that case, if you're going back to the right, aren't you removing the waypoints you just added?

Reply:
Yes, but this is necessary to remove any over-traversal when moving up to a row with fewer nodes. If we do not remove those nodes, the graph will stick out like this
<img width="739" alt="Screen Shot 2022-11-13 at 7 50 49 PM" src="https://user-images.githubusercontent.com/56476240/201554190-2e7fed62-8f53-4e8a-bfe5-dfa80d108c1a.png">